### DISCUSSION TITLE
Use cases for reentry

### DISCUSSION THREAD

**Discussion by leighmcculloch** [stellar/stellar-protocol#1594] [Core Advancement Proposals]
Hi all! There's been some discussion in Discord and elsewhere about reentry. It'd be helpful to have a good understanding of how reentry would be used if the feature was added.

Please post a new top-level comment in this discussion for each use case reentry would have. Be as detailed as possible in explaining why reentry is needed for that use case.

This discussion thread is not about any specific reentry proposal. For reentry proposals, please start a new discussion. This discussion thread is just a place to collect use cases.

**Comment by kalepail** [comment]
I'm working on smart wallets and we have the concept of policy signers which are separate contracts which execute as a cross contract call from within the smart wallet's `__check_auth` method. It would be nice from within the policy to be able to call back to the smart wallet to read in at least get methods for reading contract data like signer info if not also being able to do full writes for modifying the smart wallet itself. Currently this isn't possible and the only real work around is passing a lot of info in via arguments which can make it difficult then to filter and operate conditionally on. 

And there's no work around for modifying the smart wallet itself from within the policy during a policy invocation from the smart wallet

**Reply by leighmcculloch** [reply]
For the read only data requests could the wallet make two calls to the policy signer, first asking for a list of data, then making the final call with the accompanying data?

Doing the two steps may in some cases be more efficient than calling back N times. Unless the data required is dependent on the values returned.

**Reply by leighmcculloch** [reply]
For the modifications to the policy could the policy pass them back in the return value?

**Reply by kalepail** [reply]
> could the wallet make two calls to the policy signer

Hmm...maybe? I hadn't thought about this route. But probably?? It's backwards to what you'd expect to do but likely a sufficient work around, especially given the limited and standard things a policy could potentially request from a wallet

> for the modifications to the policy could the policy pass them back in the return value?

Also possibly. Again a bit backwards to what you'd expect but could definitely work I think

I'll do some exploring. Thanks!

**Comment by jayz22** [comment]
The most typical example is one of flash loan, where a borrower calls into a liquidity pool making a loan and the pool calls the borrower contract back to perform some custom logic (before returning the loan), all within the same invocation atomically. 

Tagging @esteblock and @rudewalt who have requested it in the original tracking issue https://github.com/stellar/rs-soroban-env/issues/336

**Reply by andreas-osec** [reply]
In flash loans, the invoker doesn’t necessarily need to be the borrowing contract - in fact, it typically isn’t. In such scenarios, reentrancy isn’t required. The [xycloans implementation](https://github.com/xycloo/xycloans/blob/d9a7ae163e6c7120000532bf4b15b6a96b3a82bc/pool/src/contract.rs#L215) is good example of this approach

**Reply by brozorec** [reply]
The invoker doesn’t necessarily need to be the borrowing contract, but what if the borrowing contract needs to call the liquidity pool back. Let me give an example with the case of a collateral swap.

Assuming the borrower has an open debt position at a lending market, with BTC as collateral and USDC as debt. They deem XLM has a higher upward potential and want to swap their pledged BTC for XLM. This can be achieved with a flash loan:

0. User invokes pool's flash loan function
1. Flashloan USDC (pool calls borrower after transferring flash-loaned amount)
2. Repay USDC and Withdraw BTC (borrower calls pool)
3. Swap BTC for XLM (borrower calls 3rd contract)
4. Deposit XLM and Borrow USDC (borrower calls pool)
5. Approve USDC to repay flash loan

We can see from here the borrower contract needs to call the pool back at least 2 times to accomplish the collateral swap.
